File: java/p2/src/p2/lang/plan/Expression.java
package p2.lang.plan;

import java.util.Set;

public abstract class Expression {
	
	private xtc.tree.Location location;
	
	public void location(xtc.tree.Location location) {
		this.location = location;
	}
	
	public xtc.tree.Location location() {
		return this.location;
	}

	@Override
	public abstract String toString();
	
	/**
	 * @return The java type of the expression value.
	 */
	public abstract Class type();
	
	public abstract Set<Variable> variables();
	
}

File: lincoln/ruby/lib/lang/plan/expression.rb
class Expression
  attr_accessor :position, :location

  def to_s
    throw "Abstract method Expression.to_s not subclassed"
  end
    
  def variables
    throw "Abstract method Expression.variables not subclassed"
  end
    
  def function
    throw "Abstract method Expression.to_s not subclassed"
  end
end